README.md
[![appveyor](https://ci.appveyor.com/api/projects/status/github/DTOcean/dtocean-moorings?branch=master&svg=true)](https://ci.appveyor.com/project/DTOcean/dtocean-moorings)
[![codecov](https://codecov.io/gh/DTOcean/dtocean-moorings/branch/master/graph/badge.svg)](https://codecov.io/gh/DTOcean/dtocean-moorings)
[![Codacy Badge](https://app.codacy.com/project/badge/Grade/8deb46aea7fd41a9baa9993e9d4279da)](https://www.codacy.com/gh/DTOcean/dtocean-moorings/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=DTOcean/dtocean-moorings&amp;utm_campaign=Badge_Grade)
[![release](https://img.shields.io/github/release/DTOcean/dtocean-moorings.svg)](https://github.com/DTOcean/dtocean-moorings/releases/latest)

# DTOcean Mooring and Foundations Module

This package provides the Mooring and Foundations design module for the DTOcean 
tools. It can design the required station keeping for array of fixed or floating 
wave or tidal ocean energy converters (constrained by the environment and 
device design), and calculate the cost. It optimises the design for minimum 
cost given a specified safety factor. 

See [dtocean-app](https://github.com/DTOcean/dtocean-app) or [dtocean-core](
https://github.com/DTOcean/dtocean-app) to use this package within the DTOcean
ecosystem.

\* For python 2.7 only.

## Installation

Installation and development of dtocean-moorings uses the [Anaconda 
Distribution](https://www.anaconda.com/distribution/) (Python 2.7)

### Conda Package

To install:

```
$ conda install -c dataonlygreater dtocean-moorings
```

### Source Code

Conda can be used to install dependencies into a dedicated environment from
the source code root directory:

```
$ conda create -n _dtocean_moor python=2.7 pip
```

Activate the environment, then copy the `.condrc` file to store installation  
channels:

```
$ conda activate _dtocean_moor
$ copy .condarc %CONDA_PREFIX%
```

OR, if you're using Powershell:

```
$ conda activate _dtocean_moor
$ copy .condarc $env:CONDA_PREFIX
```

Install [polite](https://github.com/DTOcean/polite) into the environment. For 
example, if installing it from source:

```
$ cd \\path\\to\\polite
$ conda install --file requirements-conda-dev.txt
$ pip install -e .
```

Finally, install dtocean-moorings and its dependencies using conda and pip:

```
$ cd \\path\\to\\dtocean-moorings
$ conda install --file requirements-conda-dev.txt
$ pip install -e .
```

To deactivate the conda environment:

```
$ conda deactivate
```

### Tests

A test suite is provided with the source code that uses [pytest](
https://docs.pytest.org).

If not already active, activate the conda environment set up in the [Source 
Code](#source-code) section:

```
$ conda activate _dtocean_moor
```

Install packages required for testing to the environment (one time only):

```
$ conda install -y mock pytest pytest-cov=2.5.1 pytest-mock
```

Run the tests:

``` 
$ pytest tests
```

### Uninstall

To uninstall the conda package:

```
$ conda remove dtocean-moorings
```

To uninstall the source code and its conda environment:

```
$ conda remove --name _dtocean_moor --all
```

## Usage

Example scripts are available in the "examples" folder of the source code.

```
$ cd examples
$ python fairhead.py
```

## Component database keys and values

The keys of the `compdict` argument to the Variables class have the following
meanings:

*   item1: Mooring or foundation system
*   item2: Name
*   item3: Subname
*   item4: [Material, Grade, Colour]
*   item5: Strength and mechanical properties. For piles [yield stress, Young's 
           modulus]. For ropes [minimum break load, [load, axial stiffness]]. 
           For cables [minimum break load, minimum bend radius]. All other 
           components: [minimum break load, axial stiffness]
*   item6: Size. For piles/suction caissons [diameter, thickness]. For anchors 
           [width, depth, height, connecting size]. For chains, forerunner  
           assemblies, shackles and swivels [diameter, connecting length]. For 
           ropes and cables [outer diameter]
*   item7: Mass. For piles/suction caissons, chains, ropes, cables and 
           forerunner  assemblies [dry mass per unit length, wet mass per unit 
           length]. For anchors, shackles and swivels [dry unit mass, wet unit 
           mass]
*   item8: Environmental impact
*   item9: Anchor coefficients
*   item10: Failure rates
*   item11: Cost

## Contributing

Pull requests are welcome. For major changes, please open an issue first to
discuss what you would like to change.

See [this blog post](
https://www.dataonlygreater.com/latest/professional/2017/03/09/dtocean-development-change-management/)
for information regarding development of the DTOcean ecosystem.

Please make sure to update tests as appropriate.

## Credits

This package was initially created as part of the [EU DTOcean project](
https://www.dtoceanplus.eu/About-DTOceanPlus/History) by:

*   Sam Weller at [the University of Exeter](https://www.exeter.ac.uk/)
*   Jon Hardwick at [the University of Exeter](https://www.exeter.ac.uk/)
*   Mathew Topper at [TECNALIA](https://www.tecnalia.com)

It is now maintained by Mathew Topper at [Data Only Greater](
https://www.dataonlygreater.com/).

## License

[GPL-3.0](https://choosealicense.com/licenses/gpl-3.0/)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [2.1.0] - 2021-10-13

### Added

-   Added input option to use the maximum turbine thrust coefficient in all 
    calculations regardless of current velocity.

### Fixed

-   Fixed issue with mooring solution for first floating device being different
    from others by repeating the mooring design algorithm for the first device
    with final component set.

## [2.0.0] - 2019-03-06

### Added

-   Raise a RuntimeError if the calculated device draft differs from the given 
    equilibrium draft by more than 5%.
-   Return human readable error if anchors are not applicable to a soil type.

### Changed

-   Split core module into 5 separate modules, one for each class.
-   Broke down Loads.gpnearloc into smaller functions and fixed various bugs.
-   Fix use of gpnear coordinate for finding seabed slope in Foundations class.
-   Catch slope angle exceeding friction angle for gravity or shallow
    foundations.
-   Improved calculation efficiency by reducing number of numpy function calls.
-   Changed function of prefound argument to define a preferred foundation type
    rather than a single type. If the preferred option can not be found then the
    cheapest alternative is used.

### Removed

-   Removed non-working islay and shetland example files and data.

### Fixed

-   Fixed issues with inheritance in Main and Umb classes.
-   Fixed issue with groutpilebondstr attribute not being initialised in Found
    class.
-   Ensured drag anchor database ids are added to the bill of materials, rather
    than just "drag".
-   Fixed bug in vertical distance requirements for moorings lines.
-   Fixed bug in device draft calculation.
-   Fixed issue where the global system position was incorrectly using positive
    depth.
-   Fixed functionality for using a single design for all foundations of a 
    device("uniary").
-   Fixed bug with calculation of umbilical top connection point.
-   Fixed bug in device position retuned by the mooreqav method.
-   Fixed ULS and ALS displacement testing logic.
-   Fixed bug with non-float arguments to numpy.linalg.lstsq function.

## [1.0.0] -   2017-01-05

### Added

-   Initial import of dtocean-gui from SETIS.




examples/fairhead.py

import pytest

import os
import time
import pprint
import cPickle as pickle
import numpy as np
import pandas as pd

from dtocean_moorings import start_logging
from dtocean_moorings.main import Variables, Main

this_dir = os.path.dirname(os.path.realpath(__file__))
data_dir = os.path.join(this_dir, "..", "sample_data")

def main():    
    
    """ Fairhead dummy scenario with 10x fixed tidal turbines loosely based on the HS1000 
        turbine and 1 substation. Environmental conditions based on the Metocean Assessment of the 
        Fair Head Tidal Energy Project report by DNV GL. Note: ***Denotes information not available. """
    
    soil_grid = np.genfromtxt(os.path.join(data_dir, 'fairheadsoil.txt'),
                              dtype=(float, float, 'U2', float, 'U2', float),
                              missing_values='INFINITE',
                              filling_values=np.inf)
    soil_grid = soil_grid.tolist()
    
    # Substations parameters
    substparams = pd.read_csv(os.path.join(data_dir,
                                           'substparams_fairhead.txt'),
                              sep='\t',
                              index_col=0,
                              header=0)
    
    input_variables = Variables(['device001', 'device002', 'device003', 'device004', 'device005', 'device006', 'device007', 'device008', 'device009', 'device010'], #device list
                                9.80665, #gravity
                                1025.0, #sea water density
                                1.226, #air density
                                7851.1, #steel density
                                2400.0, #concrete density
                                2450.0, #grout density
                                eval(open(os.path.join(data_dir, 'dummycompdb.txt')).read()), #component database
                                soil_grid,
                                [], #seafloor friction coefficient (optional)
                                np.loadtxt(os.path.join(data_dir, 'fairheadbath.txt'), delimiter="\t"), #bathymetry grid
                                5.0, #grid deltax
                                5.0, #grid deltay
                                0.7, #***water level maximum offset (absolute)
                                0.81, #***water level minimum offset (absolute)
                                3.5, #current velocity
                                302.0, #***current direction
                                "1/7 power law", #current profile alternatives: "uniform" "1/7 power law"
                                [0.0, 30.0, 60.0, 90.0, 120.0, 150.0, 180.0, 210.0, 240.0, 270.0, 300.0, 330.0], #***wave direction. Leave the square bracket blank if there are no wave conditions to analyse
                                [4.13, 2.93, 3.11, 2.66, 4.79, 2.08, 1.96, 1.73, 1.87, 1.95, 2.7, 4.01], #significant wave height. Leave the square bracket blank if there are no wave conditions to analyse
                                [11.29, 9.37, 9.8, 8.92, 12.16, 7.74, 7.64, 7.18, 7.46, 7.63, 8.98, 11.12], #***peak wave period. Leave the square bracket blank if there are no wave conditions to analyse
                                [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0], #***jonswap gamma. Leave the square bracket blank if there are no wave conditions to analyse
                                32.9, #wind velocity *Note: 100m above water level
                                270.0, #***wind direction
                                42.77, #***wind gust velocity
                                270.0, #***wind gust direction
                                pd.read_csv(os.path.join(data_dir, 'soilprops.txt'), sep='\t', index_col = 0, header = 0), #default soil properties                                
                                np.loadtxt(os.path.join(data_dir, 'linebcf.txt'), delimiter="\t"), #buried line bearing capacity factors
                                np.loadtxt(os.path.join(data_dir, 'subgradereactioncoefficientk1_cohesive.txt'), delimiter="\t"), #subgrade reaction coefficients
                                3.0, #soil sensitivity
                                np.loadtxt(os.path.join(data_dir, 'subgradereactioncoefficient_cohesionless.txt'), delimiter="\t") , #subgrade soil reaction coefficients cohesionless
                                np.loadtxt(os.path.join(data_dir, 'piledeflectioncoefficients.txt'), delimiter="\t") , #pile deflection coefficients
                                np.loadtxt(os.path.join(data_dir, 'pilemomentcoefficientsam.txt'), delimiter="\t") , #pile moment coefficients am
                                np.loadtxt(os.path.join(data_dir, 'pilemomentcoefficientsbm.txt'), delimiter="\t") , #pile moment coefficients bm
                                np.loadtxt(os.path.join(data_dir, 'pilelimitingvaluesnoncalcareous.txt'), delimiter="\t"), #pile limiting values non calcaeous soils
                                np.loadtxt(os.path.join(data_dir, 'holdingcapacityfactorsplateanchors.txt'), delimiter="\t"), #plate anchor holding capacity factors
                                'tidefixed', #device type options: 'tidefloat', 'tidefixed', 'wavefloat', 'wavefixed'
                                False, #depth variation permitted
                                "cylindrical",   #device profile options: "cylindrical" "rectangular"
                                310.0e3, #device mass
                                [0.0, 0.0, 11.0], #device centre of gravity. Note for both fixed and floating devices the device origin is assumed to be at the base of the device
                                250.59, #device displaced volume
                                33.5, #device height
                                15.0, #device width
                                22.0, #device length
                                0.9e-2, #device surface roughness
                                {'device001': [683000.0, 6123500.0,0.0], 'device002': [683500.0, 6123500.0,0.0], 'device003': [684000.0, 6123500.0,0.0], 'device004': [684500.0, 6123500.0,0.0], 'device005': [685000.0, 6123500.0,0.0], 'device006': [683000.0, 6124500.0,0.0], 'device007': [683500.0, 6124500.0,0.0], 'device008': [684000.0, 6124500.0,0.0], 'device009': [684500.0, 6124500.0,0.0], 'device010': [685000.0, 6124500.0,0.0]}, #device global locations. Note for both fixed and floating devices the device origin is assumed to be at the base of the device
                                np.array([[0.0,1.0,-1.5],[1.0,0.0,-1.5],[0.0,-1.0,-1.5],[-1.0,0.0,-1.5]]), #fairlead locations (from device origin)
                                np.array([[0.0, 13.333, 0],[7.5, -6.667, 0],[-7.5, -6.667, 0]]), #foundation locations (from device origin)
                                [0.0, 0.0, 22.0], #device umbilical connection point. Note for both fixed and floating devices the device origin is assumed to be at the base of the device
                                0.0, #device dry frontal area
                                0.0, #device dry beam area
                                np.loadtxt(os.path.join(data_dir, 'dragcoefcyl.txt'), delimiter="\t"), #cylinder drag coefficients
                                np.loadtxt(os.path.join(data_dir, 'wakeamplificationfactorcyl.txt'), delimiter="\t"), #cylinder wake amplification factors
                                np.loadtxt(os.path.join(data_dir, 'winddragcoefrect.txt'), delimiter="\t"), #rectangular wind drag coefficients
                                np.loadtxt(os.path.join(data_dir, 'currentdragcoeffrect.txt'), delimiter="\t"), #rectangular current drag coefficients
                                np.loadtxt(os.path.join(data_dir, 'driftcoefficientfloatrect.txt'), delimiter="\t"), #rectangular wave drift coefficients
                                (23.0, 0.0), #rotor diameter and hub offset from system centreline
                                np.loadtxt(os.path.join(data_dir, 'thrustcurv.txt'), delimiter="\t"), #thrust curve
                                22.0, #hub height. The vertical height from the system origin (fixed devices: seafloor, floating devices: base of device)
                                0.0, #device orientation angle
                                [],#pickle.load(open(str(data_dir)+'\\fex_wp4.pkl','rb')), #first-order wave excitation forces with analysed frequencies and directions
                                'catenary', #predefined mooring system type options: 'catenary', 'taut'
                                [20.0, 20.0, 10.0], #device maximum displacements in surge, sway and heave
                                'uniarypile', #predefined foundation type options: 'shallowfoundation', 'gravity', 'pile', 'suctioncaisson', 'directembedment', 'drag'
                                1.0, #steel cost
                                0.25, #grout cost
                                0.24, #concrete cost
                                125.0, #grout strength
                                'id742', #predefined umbilical type (database identification number)
                                1.4925, #umbilical safety factor from DNV-RP-F401
                                1.5, #foundation safety factor
                                [], #predefined footprint radius
                                {'device001': [683002.0, 6123502.0,-60.0], 'device002': [683502.0, 6123502.0,-60.0], 'device003': [684002.0, 6123502.0,-60.0], 'device004': [684502.0, 6123502.0,-60.0], 'device005': [685002.0, 6123502.0,-60.0], 'device006': [683002.0, 6124502.0,-60.0], 'device007': [683502.0, 6124502.0,-60.0], 'device008': [684002.0, 6124502.0,-60.0], 'device009': [684502.0, 6124502.0,-60.0], 'device010': [685002.0, 6124502.0,-60.0]}, #global subsea cable connection point
                                substparams,
                                1.7, #mooring ultimate limit state safety factor from DNV-OS-E301
                                1.1, #mooring accident limit state safety factor from DNV-OS-E301
                                6.0, #grout safety factor
                                98.6, #device wet frontal area
                                132.5, #device wet beam area
                                0.0, #device equilibrium draft without mooring system
                                np.loadtxt(os.path.join(data_dir, 'waveinertiacoefrect.txt'), delimiter="\t"), #wave inertia coefficient for rectangular structures
                                [], #predefined mooring line component list e.g. ['shackle001','rope','shackle002']
                                1.0) #optional fabrication cost factor
    test = Main(input_variables)    
    devices = test()
    
    return devices
    
# def plot(rsystime):
    
    # plt.plot(rsystime)
    # plt.ylabel('System reliability', fontsize=10)
    # plt.xlabel('Time [hours]', fontsize=10)     
    # plt.show()
    
if __name__ == "__main__":

    start_logging(level="DEBUG")
    
    devices = main()
    
    pprint.pprint(devices)
    
    # plot(rsystime)

